### PR TITLE
Final changes for mfem-4.7

### PR DESCRIPTION
**Pull request by tzanio**
📆 **Target release date:** April 12, 2024

🟡 **In Progress:**

🔴 **TODO:**

🟢 **DONE:**
- [x] #3648
- [x] #4156 
- [x] #4084 
- [x] #4067 
- [x] #4187 
- [x] #4173 
- [x] #4198 
- [x] #4121
- [x] `schwarz_ex1` does not check the `visualization` variable when sending data to GLVis
- [x] The `nurbs_patch_ex1` tests defined in the CMake build system do not turn off visualization
- [x] #4199
- [x] Clean `*.dSYM` files in the `config` directory, e.g. from building `get_hypre_version`
- [x] In `miniapps/spde/CMakeLists.txt`, get rid of the message when MPI is not enabled. Add a test for the miniapp, similar to the one in the GNU make build system. [Added in #3922]
- [x] #3922
- [x] #4211 
- [x] #4151
- [x] #4015
- [x] #4056
- [x] #3598
- [x] #4044
- [x] #4226 
- [x] #4113
- [x] #4135
- [x] #4227
- [x] #4225 
- [x] #4203
- [x] #4054
- [x] #4206
- [x] #4236
- [x] #4128 
- [x] Add suggestion from https://github.com/orgs/mfem/discussions/4207#discussioncomment-8946015
- [x] Add this suggestion: https://github.com/mfem/mfem/pull/4151#discussion_r1542101656
- [x] Fix MUMPS build with single precision, see: https://github.com/mfem/mfem/pull/3922#issuecomment-2023779007; also merge this fix: #4214
- [x] #3904
- [x] #4243
- [x] #1951
- [x] #4274
- [x] #2669
- [x] #4276
- [x] #4253
- [x] #3844
- [x] Update `CHANGELOG` ← @tzanio ; :warning: There are some edits in `CHANGELOG` in `master` compared to `v4.6` in older release sections -- these need to be moved to the section for the new release.
- [x] Ask MFEM-based applications to test the `mfem-4.7-dev` branch ← @tzanio
- [x] Website PR: https://github.com/mfem/web/pull/263

🟡 **Postponed:**
- [ ] #4175
- [ ] Fix the following error messages from `ex1p` with "mfem+cuda" and "hypre+cuda" build:
   ```console
   CUDA ERROR (code = 4, driver shutting down) at device_utils.c:148
   CUDA ERROR (code = 4, driver shutting down) at general.c:255
   ```

### CODE REVIEW

**Comment by artv3** [comment]
When I build with my application I am running into a build issue: 
```
include/mfem/linalg/../general/forall.hpp:225:29: error: no member named 'cuda_launch_t' in namespace 'RAJA'
   RAJA::LaunchPolicy<RAJA::cuda_launch_t<true>>;
```

**Comment by stefanozampini** [comment]
@tzanio Please add https://github.com/mfem/mfem/pull/4203 to the list of PRs. It is a no brainer

File: CHANGELOG [hunk 1]
```diff
@@ -19,61 +29,57 @@ Meshing improvements
...
   advection, Burgers', shallow-water, Euler equations (see, Example 18) are
   available.
 
-GPU support
-----------------------------
+- Added functionality for construction of cut-surface and cut-volume
+  IntegrationRules through a moment-fitting approach. The cut is specified by
+  the zero level set of a Coefficient. See fem/intrules_cut.hpp and Example 38.
+
+- Introduced support for higher order nonconformal Nedelec elements on simplicesa
```
Comment (dylan-copeland) [inline CHANGELOG:43]: Typo

**Comment by dylan-copeland** [comment]
We need this bug fix for single builds: https://github.com/mfem/mfem/pull/4235

**Comment by tzanio** [comment]
@v-dobrev, how do you prefer to resolve the `makefile` conflict (due to https://github.com/mfem/mfem/pull/4214/files#diff-0c53bce8f74b4f469712caf7ce78af7f020c9700387c4739732f8acf09517fd9R173-R184 and https://github.com/mfem/mfem/pull/4172/commits/18cf9d7ea16981835b9036b3d72ebd91bd0d178c)?

File: config/config.hpp [hunk 2]
```diff
@@ -120,6 +120,15 @@ constexpr real_t operator""_r(unsigned long long v)
...
 // Check dependencies:
 
+// Define MFEM_MPI_REAL_T to be the appropriate MPI real type
+#ifdef MFEM_USE_MPI
+#ifdef MFEM_USE_SINGLE
+#define MFEM_MPI_REAL_T MPI_FLOAT
+#elif defined MFEM_USE_DOUBLE
+#define MFEM_MPI_REAL_T MPI_DOUBLE
+#endif
+#endif                                          \
```
Comment (v-dobrev) [inline config/config.hpp:130]:
```suggestion
#endif
```

**Comment by v-dobrev** [comment]
> @v-dobrev, how do you prefer to resolve the `makefile` conflict (due to https://github.com/mfem/mfem/pull/4214/files#diff-0c53bce8f74b4f469712caf7ce78af7f020c9700387c4739732f8acf09517fd9R173-R184 and [18cf9d7](https://github.com/mfem/mfem/commit/18cf9d7ea16981835b9036b3d72ebd91bd0d178c))?

There's really no conflict there but git does not quite understand that -- the first part was moved to `defaults.mk` and the second part was added here. Maybe the proximity of the two is confusing for git.

**Comment by v-dobrev** [comment]
@tzanio, I merged `master` here but some of the edits in CHANGELOG made it difficult to figure out what's new and what's been editted here, so I kept both versions of the conflict; to remind us about this, I added this note to the TODO bullet about the CHANGELOG:

> - [ ] Update `CHANGELOG` ← @tzanio. ⚠️  Due to re-merging of `master` some CHANGELOG entries may have been re-introduced here after they were modified in this branch.

**Comment by tzanio** [comment]
> @tzanio, I merged `master` here but some of the edits in CHANGELOG made it difficult to figure out what's new and what's been editted here, so I kept both versions of the conflict; to remind us about this, I added this note to the TODO bullet about the CHANGELOG:
> 
> > * [ ]  Update `CHANGELOG` ← @tzanio. ⚠️  Due to re-merging of `master` some CHANGELOG entries may have been re-introduced here after they were modified in this branch.

I updated `CHANGELOG` and resolved the conflicts

**Comment by najlkin** [comment]
As revealed in #4262 , `minimal-surface.cpp` does not build with single precision 😱

**Comment by najlkin** [comment]
Perhaps we could put the commit (2e37f2ccb49eb6fcb48917c7e6fc194088588abb) fixing `(p)minamal-surface.cpp` here? 🤔

**Comment by v-dobrev** [comment]
> Perhaps we could put the commit ([2e37f2c](https://github.com/mfem/mfem/commit/2e37f2ccb49eb6fcb48917c7e6fc194088588abb)) fixing `(p)minamal-surface.cpp` here? 🤔

Sounds good. Can you please push the change here?

**Comment by tzanio** [comment]
This branch is now a release candidate for `mfem-4.7`.

Please test it in your application code and let us know if you discover any issues.

ping: @artv3, @tomstitt, @btalamini, @jamiebramwell, @sethwatts, @sebastiangrimberg, @sshiraiwa

**Review by mlstowell — APPROVED** [review]

**Review by v-dobrev — APPROVED** [review]

**Review by najlkin — APPROVED** [review]

**Pull request merged**